### ISSUE TITLE
Timecode Issue for Elementary Streams

### ISSUE THREAD

**Issue by Royal-star** [mbunkus/mkvtoolnix#1564]
Dear Mr. Bunkus!
For the first, please sorry for my poor English, but I will try to be clear in my expression.
Like the last time I had report to you for the stocking of HD Footage after merging it with MkvMerge.
I have in the meanwhile discover the Problem and find out an very important lac of mkv Merge.
So it appears that after converting of an HD Video Footage and then following with an Merge of Converted Video and Source DTS Audio File with MkvToolNix, the Resulting MKV output appears with Elementary Stream Time code Issues (It is not Playable). I have tested it out with mkclean and the other fix Tools eg.  mkclean.exe.
So my Request is for this awesome Work of you.
Can it be that during the Mux Process the Timecodes of the involved Elementary Streams
Compared and new created like in TsMuxer, so I do not need to remux it with TS Muxer to a m2ts File and then again merge it with MKVTooLnix to a MKV File to become a playable
Mkv File.
Thank you in advice for your hard working and spending your Time for hearing us.
SAMARAS DIMITRIOS

**Comment by mbunkus** [comment]
I honestly don't understand what your issue is; I'm sorry. Can you please upload a file to my [FTP server](/mbunkus/mkvtoolnix/wiki/FTP-server) before muxing and then tell me exactly what you think mkvmerge is doing wrong?

**Comment by Royal-star** [comment]
Thank you for the prompted answer.
Upon using your remux program I had the following issue:
After remuxing a video file from a blueray .m2ts I got an .mkv which contained a .avc a .dts  and a .sup file that played perfectly. Then I wanted to reduce the size of the .mkv (it was 20 Gb). I extracted the .avc with the gMKVextractGUI program and then I converted it to a smaller .avc. Now, when I remuxed the new .avc with the originals .dts and .sup files I created a new, smaller in size .MKV. When I played the new .MKV (on numerous players) after a few seconds it freezes completely. Upon personal research I discovered that after conversion of the .avc the time codes are different in comparison to the original elementary streams of the source container, resulting to this fatal error. So in order to have my new smaller .MKV container perfect again, I had to take those three elementary files (new .avc and original .dts .sup) and make a .m2TS container with the TSmuxer program (because it synchronizes the timecodes) and then finally I put the .m2TS through your program to make the .MKV that I wanted. What I ask is if you can make your program to correct the time codes, like the TSmuxer program does, of the elementary files when it creates the .MKV
Hopefully I made a better description of the problem and if you need to ask me something else I am at your disposal.

**Comment by Royal-star** [comment]
Thank you for the prompted answer.
Upon using your remux program I had the following issue:
After remuxing a video file from a blueray .m2ts I got an .mkv which contained a .avc a .dts  and a .sup file that played perfectly. Then I wanted to reduce the size of the .mkv (it was 20 Gb). I extracted the .avc with the gMKVextractGUI program and then I converted it to a smaller .avc. Now, when I remuxed the new .avc with the originals .dts and .sup files I created a new, smaller in size .MKV. When I played the new .MKV (on numerous players) after a few seconds it freezes completely. Upon personal research I discovered that after conversion of the .avc the time codes are different in comparison to the original elementary streams of the source container, resulting to this fatal error. So in order to have my new smaller .MKV container perfect again, I had to take those three elementary files (new .avc and original .dts .sup) and make a .m2TS container with the TSmuxer program (because it synchronizes the timecodes) and then finally I put the .m2TS through your program to make the .MKV that I wanted. What I ask is if you can make your program to correct the time codes, like the TSmuxer program does, of the elementary files when it creates the .MKV
Hopefully I made a better description of the problem and if you need to ask me something else I am at your disposal.

**Comment by mbunkus** [comment]
Thanks for the explanation, but that's still much too vague for me to say anything about it. If you want me to look into it I'll need a couple of sample files:

- the Matroska file before you've extracted the AVC track from it
- the raw AVC track after re-encoding
- the M2TS file you've created with that re-encoded AVC track

Additionally I'll need concrete descriptions how you did each step: which programs where involved? Which settings exactly where used in each step? etc.

Please upload such files to [my FTP server](/mbunkus/mkvtoolnix/wiki/FTP-server). Thanks.

**Comment by Royal-star** [comment]
First I remux with MKVToolNix, on default setting, a .avc, a .DTS and a .Sup File from BluRay M2ts Stream to a .MKV (The Matroska file before i've extracted the .AVC track from it).
Then I extract from this .mkv all three Files with gMKVExtractGUI to Elementary Streams (.avc, .DTS, .sub).
I converted the .avc with TotalCodeStudio (Settings:High 4.2,VBR-15000kbps,RefFr 3,B-fr=2,NTSC 23.976,16:9,1920x1080p) to x264 Raw Video File, the sample in folder is named “The raw .AVC track after re-encoding.264”.
I mux with MKVToolNix, on default setting, the converted .avc Video File with the untouched .DTS Audio Source File to a new .Mkv.
The resulting .Mkv File is not playable! The sample in folder is named “The Errorfile after mux the raw re-encodet avc +Source DTS audio.mkv”.
Then I took the same input Files (converted .avc and .Dts Source file) and make an .m2ts File with TSmuxer (on default settings), the sample  in folder is named “The M2TS file I’ve created with that re-encoded .AVC track+Source .DTS audio.m2ts”.
I took the resulting .m2ts as input for MKVToolNix and remux (on default settings) to an .mkv.
This time the resulting .mkv plays without any problems. The sample in the folder is named “The final playable mkv with the raw re-encodet avc +Source DTS audio from the m2ts File”.
As before, my hope is that I am specific enough but non the less, if you have any questions or there is something else that you want me to do, I am at your disposal.
Here are the File Names of the Files I send to you.
The Errorfile after mux the raw re-encodet avc +Source DTS audio.mkv
The final playable mkv with the raw re-encodet avc +Source DTS audio from the m2ts File.mkv
The M2TS file i've created with that re-encoded AVC track+Source DTS audio.m2ts
The Matroska file before i've extracted the AVC track from it.mkv
The raw AVC track after re-encoding.264

**Comment by mbunkus** [comment]
Thanks, I'll look into it. I cannot say when, exactly, probably around the weekend.

**Comment by Royal-star** [comment]
Thank you very much Mr Bunkus and take your time for make this world better.

**Comment by mbunkus** [comment]
I'm sorry, but I fail to see the problem with the files you've uploaded. You said that the file created from the re-encoded h.264 track doesn't work. That's the file `The Errorfile after mux the raw re-encodet avc +Source DTS audio.mkv`, correct? That one plays just fine here on both VLC and MPC-HC.

The file is completely spec-compliant. It's even fully in sync.

One notable difference between this file and the final file you say is working for you, `The final playable mkv with the raw re-encodet avc +Source DTS audio from the m2ts File.mkv`, is that the the frames in `The Errorfile…` are all three bytes longer. I don't know why that is, but it makes no difference to the decoding.

I've verified this by extracting the first 100 frames from both files with ffmpeg into 100 PNG files each. Then I've created SHA-256 checksums of all the PNGs and compared them. They were identical.

So for all intents and purposes `The Errorfile…` is OK. If your player doesn't like the file then it's not mkvmerge's fault, and I cannot help you. I'm sorry.

**Comment by mbunkus** [comment]
Alright, I see the problem now… There are several seconds of video frames simply missing. I'll investigate further.

**Comment by Royal-star** [comment]
Let me add some additional information please, the problem occurs (at least in my players, please, and I can’t put enough stress on it, in all my players) after a minimum duration of at least nine (9) seconds. I have checked it again and again. The Error-File stuck exactly after 9 seconds. There is also the possibility that could take even more time in your player for reasons that escapes me, after all I’m not a programmer as you are.
To specify, I have run the file in MPC-HC, VLC and Potplayer. I say again, after nine (9) seconds, it stucks. How is it possible that only happens to me.
That means that the Problem occurs beyond the 200 Frame-Mark after Start.
The 100 Frames sample that you have taken for testing was a bit to small to witness the error for your self.
I also want to add that I have the K-Lite Codeck Pack which I keep it updated as often as possible.
I have uploaded a Frame capture of how I see it when it stucks to your FTP Server (The Errorfile after mux the raw re-encodet avc +Source DTS audio.mkv_20160117_203208.526.jpg).
Please note, because I think that I get a hint of irony in the end of your last response, that I’m just a simple user and I prefer your program for various reasons of my own, but please think a little more highly of me that I wouldn’t try to waste yours and my time as well completely pointlessly if there wasn’t an actual flaw to deal with.
So, again, thank you for your time and if you need more details or information, as always, I’m at your disposal.

**Comment by mbunkus** [comment]
I cannot reproduce this issue with v8.8.0, neither on Linux nor on Windows. I've tried to re-create what you've done according to your description:

1. I've extracted the DTS audio track from your uploaded file `The Matroska file before i've extracted the AVC track from it.mkv`.
2. Next I've re-muxed the h.264 file you've uploaded (`The raw AVC track after re-encoding.264`) together with the DTS file extracted in step 1 with v8.8.0 and default settings into a new Matroska file called `linux.mkv` respectively `windows.mkv`.
3. Afterwards I've looked at the timestamps of the video track in both `linux.mkv` and `windows.mkv`. They're continuous – there's no gap as there is in the file you've uploaded, `The Errorfile after mux the raw re-encodet avc +Source DTS audio.mkv`.
4. Finally I've watched both files with VLC from start to finish. Watching your file, `The Errorfile after mux the raw re-encodet avc +Source DTS audio.mkv`, in VLC I can see the problem after 9 seconds quite clearly (the video stops for ~7 seconds and continues with a distortion in the picture). However, the two files I've created show no such gap, neither at the 9 seconds mark nor anywhere else.

So I don't see that there's actually a real problem (anymore) – it may also be a hardware or software issue on your end. If I cannot reproduce it then there's nothing I can do about it either.

About the irony you think you detected: I'm sorry it came across that way; it definitely wasn't intended as such. When I wrote that reply I only watched the first five seconds or so of the file, and after that I did that ffmpeg test – again with only roughly ~100 frames which amounts to roughly three seconds of video. Therefore I simply hadn't stumbled about the issue.

**Comment by Royal-star** [comment]
Thank you so much for your time, the conclusion seems to place the flaw definitely at my end, sorry for the extra am mound of work, because during the first steps of my search it seemed reasonable to be a flaw of the program it self. I'm beginning to have a greater understanding though of the source of the problem and hopefully I will locate it very soon. Non the less, if by any chance locate it back to your program, rest assured that i will keep you inform about it. On a personal note, i must say that I admire your work and that I can fully understand how easy is for a misinterpretation of a phrase to take place during a written conversation so don't worry, there was no offense taken by my end. Again, thank you for your professional approach of the matter and wish you all the best. Keep up the good work.

**Comment by mbunkus** [comment]
You're quite welcome. I'm going to close this issue for the time being.

If you manage to reproduce this on a consistent basis (e.g. when you mux the same set of files multiple times to different output files all output files will show the same kind of missing frames/gap) then post here and I'll re-open it.

**Comment by Royal-star** [comment]
Mr mbunkus, It's been a long time. I'm in the pleasant position to inform you that I have, through my research regarding the "flaw" that has troubled my project's, that finally has been a breakthrough. As we concluded last time, it was indeed an error that occurred in my back end. Normally I wouldn't have write something but I believe that it will interest you. The whole problem was generated from a simple setting on my computer management regarding my external HDD's and the way they were connected to my PC. On the properties window of each HDD, in the sub label "Policies" there are two options about the removal policy of the USB device. One option is "Quick removal" and the second is "Better performance". Not to get to technical, as default option was always the quick removal option, but, as it turned out to be, it should have been the better performance. The reason I'm telling you all these is because I have found out that other muxing programs have the advantage of verify and correct any errors that may be created during the data transfer from secondary HDD to the destination files. And this is the reason why I had this "flaw" only with the use of your program. Thank you very much, for with out your involvement it would have been highly unlikely to determine the cause of this error. As usual, if I can help or have any questions for me, I am at your disposal.
